src/librustc/mir/mod.rs

```diff
@@ -69,13 +69,38 @@ impl<'tcx> HasLocalDecls<'tcx> for Mir<'tcx> {
     }
 }
 
+/// The various "big phases" that MIR goes through.
+///
+/// Warning: ordering of variants is significant
+#[derive(Copy, Clone, RustcEncodable, RustcDecodable, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum MirPhase {
+    Build,
+    Const,
+    Validated,
+    Optimized,
+}
+
+impl MirPhase {
+    /// Gets the index of the current MirPhase within the set of all MirPhases.
+    pub fn phase_index(&self) -> usize {
+        match self {
+            MirPhase::Build => 0,
+            MirPhase::Const => 1,
+            MirPhase::Validated => 2,
+            MirPhase::Optimized => 3,
+        }
+    }
+}
+
 /// Lowered representation of a single function.
 #[derive(Clone, RustcEncodable, RustcDecodable, Debug)]
 pub struct Mir<'tcx> {
     /// List of basic blocks. References to basic block use a newtyped index type `BasicBlock`
     /// that indexes into this vector.
     basic_blocks: IndexVec<BasicBlock, BasicBlockData<'tcx>>,
 
+    pub phase: MirPhase,
```

Comment:
It's probably a good idea to add a doc comment here. Something like:

---

Records how far through the "desugaring and optimization" process this particular MIR has traversed.

This is particularly useful when inlining, since in that context we instantiate the promoted constants and add them to our promoted vector -- but those promoted items have already been optimized, whereas ours have not. This field allows us to see the difference and forego optimization on the inlined promoted items.